### DISCUSSION TITLE
Geometry Service run_script_file supported API Versions?

### DISCUSSION THREAD

**Discussion by dkowollik** [ansys/pyansys-geometry#1389] [Q&A] [answered]
Hi @RobPasMue,
I am using `launch_modeler_with_geometry_service` to open files. Now I want to run SpaceClaim python script files. I have a python script that manipulates an opened active model. The script is written in API version V19 and ran without any problems in Ansys SpaceClaim 2024R2. I only received a message that the API version is old and won't be supported in future releases. 
I used the following run script line and received an error. 
```python
modeler_gs.run_discovery_script_file(file_path=r'my_script.py', import_design=True)
GeometryRuntimeError: Script run failed: The collection is empty.
```

Where do you run the script, in Discovery API or SpaceClaim API? 
What is the oldest SpaceClaim API Version you support through the GeometryService?

**Comment by RobPasMue** [comment]
Hi @dkowollik!

To answer your question, the script is ran in the Geometry Service itself. The API version should be the same for all anyway: Discovery, SpaceClaim and Geometry Service (from what I know)

Now, the problem you might be facing is that the version you are trying to use is too old for the Geometry Service. The Geometry Service has only been provided since 23R2 as part of the Ansys installation and, unless I'm mistaken, only scripts with V23 or later are supported.

By default the Geometry Service also uses the API version that is aligned with itself (i.e. V24 if you are using Ansys 24R2)

We have a test in which we are running the Geometry Service and testing a SpaceClaim API script here:

https://github.com/ansys/pyansys-geometry/blob/719d21d50dd2d3e13480779e6439c569a0de5570/tests/integration/test_runscript.py#L37-L48

The script we are sending over to the Geometry Service is found here:

https://github.com/ansys/pyansys-geometry/blob/719d21d50dd2d3e13480779e6439c569a0de5570/tests/integration/files/disco_scripts/simple_script.py#L1-L13

I would recommend you also specify the full path to the script - rather than just its relative path.

Nonetheless, we need to improve things there - better error outputs at least!!

**Reply by dkowollik** [reply]
Hi @RobPasMue! Unfortunately, I guess I have to debug my 1000 lines of code to upgrade to the new API version. Do you have upgrade scripts that do that automatically, like python 2 to 3 converters?

**Reply by RobPasMue** [reply]
Hi @b-matteo! Are you familiar with any product solution to do this? It could help out @dkowollik greatly!

**Comment by b-matteo** [comment]
Hi @RobPasMue, @dkowollik,
i confirm what @RobPasMue wrote, GeometryService has the same script interpreter as Ansys Discovery and Ansys SpaceClaim. That being said, since there's no way to select specific script api version like what you're able to do in the product, your script should be based on the same GeometryService's api version. 
To convert your script to the GeometryService's api version:
- Open Ansys SpaceClaim 2024 R2.
- From the "Design" tab, open the Script editor.
- Open the script you want to convert in the script editor.
- In the API version drop down menu, select the latest Api available (should be 242 for you). ![image](https://github.com/user-attachments/assets/c2e3be44-b606-491a-973c-8546afad0eb6)

- Try to run your script to double check the output is what you expect.
- Save as diiferent file in order to preserve the original script.
- Try to run the new script file using PyAnsys Geometry.

Please let us know if that fixes the issue you're having.

**Reply by dkowollik** [reply]
Hi @b-matteo, @RobPasMue,
Thank you for the quick help. I came up with the same idea this morning and gave it a try. The conversion and execution of the script with API V242 works fine, if I use the SpaceClaim GUI importer for the Rhino file with the following options (excuse me for the german GUI interface):
![grafik](https://github.com/user-attachments/assets/06785aa5-2d8d-44de-8948-84c2221bf0b6)
![grafik](https://github.com/user-attachments/assets/01d1fbe8-aa12-429b-9869-599d3e69f409)

The importer is a different one, if  I use the pyansys open_file interface. I realized this, because I had to open up a Defender port to be able to use the spaceclaim snippets below. This importer interprets the model differently, so that the script file produces errors. I used the following two variants to import into pyansys and received the same errors that the model entities could not be found:

```python
from ansys.geometry.core import launch_modeler_with_spaceclaim
modeler = launch_modeler_with_spaceclaim()
model = modeler.open_file('my_rhino_file.3dm')
modeler.run_discovery_script_file(file_path=r'my_script_API242.py', import_design=True)
```
or

```python
from ansys.geometry.core import launch_modeler_with_geometry_service
modeler = launch_modeler_with_geometry_service()
model = modeler.open_file('my_rhino_file.3dm')
modeler.run_discovery_script_file(file_path=r'my_script_API242.py', import_design=True)
```

Unfortunately, I cannot share my script and the model on this platform.  

If I open the *.scdoc file that I imported manually in the SpaceClaim GUI and saved it, it works fine to execute the script within pyansys.

```python
from ansys.geometry.core import launch_modeler_with_spaceclaim
modeler = launch_modeler_with_spaceclaim()
model = modeler.open_file('my_rhino_file_imported_to_spaceclaim_gui.scdoc')
modeler.run_discovery_script_file(file_path=r'my_script_API242.py', import_design=True)
```

My question is now, how do I import the Rhino model with the exact same importer and options within the SpaceClaim GUI directly from pyansys within ScpaceClaim and GeometryService? 

My idea was to just do the import within the IronPython script itself in order to have a quick workaround. 
I recorded the following macro in SpaceClaim GUI:
```python
# Python Script, API Version = V242
DocumentOpen.Execute(r"my_rhino_file.3dm", FileSettings1)
```
`FileSettings1` seems to be some special meta object that does not really work, if I save this in my script and execute in the GeometryService. How can I create the FileSettings1 object? Or do you have a better idea?

**Reply by b-matteo** [reply]
@dkowollik Thanks for the detailed answer. i understand that you can't share any protected model on this platform. Can you share with us a simple Rhino model with a very basic geometry that we can use to test and debug importing it through scripting? 
This would help a lot

**Reply by dkowollik** [reply]
Hi @b-matteo, @RobPasMue,
I created a simple chill area Rhino model, which shows similar behaviour on the different tree structure depending on import I use. [chill_area_rhino7.zip](https://github.com/user-attachments/files/16869224/chill_area_rhino7.zip)
The example looks like this in Rhino (chill_area_rhino7):
![grafik](https://github.com/user-attachments/assets/0ed85de0-4f02-4853-96f2-45c252dbbdbd)
I added 9 solids and one surface in a layer tree structure. I also assigned names to the solid objects and the surface object. 

In SpaceClaim GUI it looks like this with the settings chosen as above: 
![grafik](https://github.com/user-attachments/assets/95451104-3e19-41c7-a0c2-6d4230310669)
This is how we expect the import from the past (All layers as components and named solids and surfaces). Here the colors are arbitrary after the import. 

The script import to the SpaceClaim GUI via CPython generates this result: 
![grafik](https://github.com/user-attachments/assets/6ab150ad-9127-4c8a-b04b-326bd4869ae6)
Extra components are generated for the solid and surface objects with their names and objects located inside these components with default naming convention. The colors, although as pastel tones, seem similar to the Rhino model inside Rhino.

**Reply by RobPasMue** [reply]
Looks like the colors are exactly the same which is kind of nice, but we should definitely work on those imports so that the final tree looks the same.

**Reply by dkowollik** [reply]
Hi @b-matteo, @RobPasMue!
This might be important for debugging or understanding the behaviour of the Services and API's. With the simple model I tried further imports (modeler.open_file, IronPython DocumentOpen and open via GUI menu). If SpaceClaim ist launched through the PyGeometry CPython interface, the import behaviour is always identical like the third screenshot above. If I start the SpaceClaim GUI as standalone, the import behaviour for IronPython DocumentOpen and open via GUI menu shows also the same behaviour like the second screenshot above. Can you reproduce this on your system?

**Comment by RobPasMue** [comment, marked as answer]
Hi @dkowollik - following the implementation on #1395:

* Choosing the API version you want to run your script with is now possible - **only with SpaceClaim and Discovery**.
* The **Geometry Service only supports the version it belongs to** (i.e. v242 for Ansys Geometry Service 24R2 and so on).
* This feature is available starting on PyAnsys Geometry **v0.7.1 and onwards.**

See more instructions on the API reference for ``modeler.open_file()`` at https://geometry.docs.pyansys.com/version/stable/api/ansys/geometry/core/modeler/Modeler.html#Modeler.run_discovery_script_file

I believe this would serve as an answer to the main topic of discussion in the description. If so, please mark as answer for other users to benefit from it.
